picasso/src/main/java/com/squareup/picasso/UrlConnectionDownloader.java

```diff
@@ -91,10 +91,23 @@ protected HttpURLConnection openConnection(Uri path) throws IOException {
     }
 
     int responseCode = connection.getResponseCode();
-    if (responseCode >= 300) {
-      connection.disconnect();
-      throw new ResponseException(responseCode + " " + connection.getResponseMessage(),
-          networkPolicy, responseCode);
+    if (responseCode != HttpURLConnection.HTTP_OK) {
+      if (responseCode == HttpURLConnection.HTTP_MOVED_TEMP
+              || responseCode == HttpURLConnection.HTTP_MOVED_PERM
+              || responseCode == HttpURLConnection.HTTP_SEE_OTHER) {
+        // Redirect
+
+        // get redirect url from "location" header field
+        String newUrl = connection.getHeaderField("Location");
```

Comment:
Shouldn't we first call `connection.disconnect()` before reassigning this? Docs state "Releases this connection so that its resources may be either reused or closed" so it sounds like we would want to call `disconnect` before reassignment